### PR TITLE
Ruby: Desugar hash literals

### PR DESCRIPTION
**Pull request by hvitved**
```rb
{ a: 1, **splat, b: 2 }
```

becomes

```rb
::Hash.[](a: 1, **splat, b: 2)
```

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/ast/internal/Synthesis.qll [hunk 1]
```diff
@@ -794,14 +794,15 @@ private module ArrayLiteralDesugar {
...
       i = -1 and
-      child = SynthChild(MethodCallKind("[]", false, al.getNumberOfElements() + 1))
+      child = SynthChild(MethodCallKind("[]", false, al.getNumberOfElements()))
       or
-      exists(AstNode mc | mc = TMethodCallSynth(al, -1, _, _, _) |
-        parent = mc and
+      exists(AstNode mc |
+        mc = TMethodCallSynth(al, -1, _, _, _) and
+        parent = mc
+      |
```
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/ast/internal/Synthesis.qll:802]: Could this just be `parent = TMethodCallSynth(al, -1, _, _, _)`?

File: ruby/ql/lib/codeql/ruby/ast/internal/Synthesis.qll [hunk 2]
```diff
@@ -825,13 +826,58 @@ private module ArrayLiteralDesugar {
...
+  private predicate hashLiteralSynthesis(AstNode parent, int i, Child child) {
+    exists(HashLiteral hl |
+      parent = hl and
+      i = -1 and
+      child = SynthChild(MethodCallKind("[]", false, hl.getNumberOfElements()))
+      or
+      exists(AstNode mc |
+        mc = TMethodCallSynth(hl, -1, _, _, _) and
+        parent = mc
+      |
```
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/ast/internal/Synthesis.qll:847]: Similarly here, `parent = TMethodCallSynth(hl, -1, _, _, _)`?

File: ruby/ql/lib/codeql/ruby/ast/internal/Synthesis.qll [hunk 3]
```diff
@@ -825,13 +826,58 @@ private module ArrayLiteralDesugar {
     final override predicate methodCall(string name, boolean setter, int arity) {
       name = "[]" and
       setter = false and
-      arity = any(ArrayLiteral al).getNumberOfElements() + 1
+      arity = any(ArrayLiteral al).getNumberOfElements()
```
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/ast/internal/Synthesis.qll:829]: Was this just incorrect, previously?

**Review by hmac — APPROVED** [review]
Just have a few minor comments, this looks good and is nice and understandable!

I guess when we start to write flow summaries for methods involving hashes, we'll make use of the `TKeywordArgumentPosition`/`TKeywordParameter` position variants in order to specify them?

File: ruby/ql/lib/codeql/ruby/ast/internal/Synthesis.qll (hunk 3 above)
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/ast/internal/Synthesis.qll:829]: Yes.

**Review by hmac — APPROVED** [review]

**Pull request merged**